### PR TITLE
Fix #1109

### LINKED ISSUE #1109: Reader mode settings not shown after relaunching the app
- Did you search for similar issues before submitting this one?
Yes

- Description:
Reader mode settings not shown after relaunching the app

- Device (iPhone5, iPhone6s plus, iPad 3, ?):
iPhone 7+ (10.3.3)

- Brave Version:
1.4.1 (17.08.24.09)

- Steps to reproduce:
    1. Visit http://theverge.com, open an article in reader mode
    2. Change backgorund from reader mode settings and close the app from background
    3. Relaunch the app, reader mode settings is missing

- Actual Behavior
See Screenshot

- Expected Behavior
Should always show the reader mode setting

- Is this an issue with Beta build?
Yes

- Is this an issue in the currently released version?
No

- Can this issue be consistently reproduced?
Yes

- Extra QA steps:
    1. 
    2. 
    3.

- Website problems only: 
- did you check with Brave Shields down? 
- did you check in Puffin/Dolphin/iCab/Firefox (UIWebView-based browsers)? 

- Screenshot if needed:
image

- Any related issues:

### PR DESCRIPTION
**Pull request by iccub**
First open source contribution ever, feel free to bash me if I did anything wrong.

I fixed the reader mode bug. 810dac1cfd7dc1c31e6f983c7f30ced31c42aae8 fixes the problem but I also made some code improvements close to the issue. They can be discarded if you don't like it.
 
### What was wrong
Logic regarding showing reading bar was called too early.
On first launch all `TabManagerDelegate` methods are called before reader state is set(reader state is set in `didChangeReaderModeState()`).

Also there are 2 `ReaderModeBarView` in app and, one created  in `BrowserViewController`, second in `BraveURLBarView`. This causes confusion in terms what to change in app.

### How I fixed it
Toggling reader mode is done in `didChangeReaderModeState()` call. It looks better because now state and bar view logic are in the same place.

Unfortunately it required to add `#if BRAVE` directive, I don't see a good way to work around it as setting reader state is called after view rendering methods

### Other minor improvements

- Adding/removing toolbar subview replaced with show/hide - it's more simpler and slightly more performant. Now the reader toolbar is created only once, and can be toggled with `isHidden` property.
- Removed redundant calls to show reader bar - in firefox showing reading bar is done in `BrowserViewController`, in Brave now it's done in `BraveURLBarView`. To not duplicate the logic, `BrowserViewController` actions now do nothing, I didn't want to modify too much firefox code, so I used `#if !BRAVE` on those methods to make them empty, while preserving the same codebase(took code from their master branch).
- Reader bar height was in wrong place, due to two reader mode bar confusion I guess.

### Testing
Tested on on iPhone 7 and iPad Air 2, both on iOS 10.3.3, works in simulator too.

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift [hunk 1]
```diff
@@ -33,12 +33,30 @@ extension BrowserViewController {
     }
 
     func showReaderModeBar(animated: Bool) {
-        (urlBar as! BraveURLBarView).showReaderModeBar()
-        updateReaderModeBar()
+        // Brave reader mode bar is different, state logic is placed in URLBarView.updateReaderModeState
+        #if !BRAVE
```
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift]: Absolutely, no way for you to know this, but we actually no longer care about keeping FF as upstream from anything outside of the `Shared` module. Please just remove these and code inside completely.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift [hunk 2]
```diff
@@ -33,12 +33,30 @@ extension BrowserViewController {
...
+            updateReaderModeBar()
+            
+            self.updateViewConstraints()
+        #endif
     }
 
     func hideReaderModeBar(animated: Bool) {
-        (urlBar as! BraveURLBarView).hideReaderModeBar()
+        // Brave reader mode bar is different, state logic is placed in URLBarView.updateReaderModeState
+        #if !BRAVE
```
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift]: Remove

File: Client/Frontend/Browser/URLBarView.swift [hunk 3]
```diff
@@ -288,6 +288,16 @@ class URLBarView: UIView {
 
     func updateReaderModeState(_ state: ReaderModeState) {
         locationView.readerModeState = state
+        
+        // Brave uses custom reader mode toolbar attached to the bottom of URL bar,
+        // after each reader mode change the toolbar needs to be toggled
+        #if BRAVE
```
Comment (jhreis) [inline Client/Frontend/Browser/URLBarView.swift]: Keep implementation, but you can just remove the `#if` wrapper.

File: brave/src/frontend/BraveURLBarView.swift [hunk 4]
```diff
@@ -93,27 +93,21 @@ class BraveURLBarView : URLBarView {
         addSubview(tabsBarController.view)
         getApp().browserViewController.addChildViewController(tabsBarController)
         tabsBarController.didMove(toParentViewController: getApp().browserViewController)
+        
+        readerModeToolbar.isHidden = true
+        readerModeToolbar.delegate = getApp().browserViewController
+        addSubview(readerModeToolbar)
```
Comment (jhreis) [inline brave/src/frontend/BraveURLBarView.swift]: Move 3 lines to `readerModeToolbar` lazy setup

File: brave/src/frontend/BraveURLBarView.swift [hunk 5]
```diff
@@ -93,27 +93,21 @@ class BraveURLBarView : URLBarView {
...
 
     func showReaderModeBar() {
-        if readerModeToolbar != nil {
-            return
-        }
-        readerModeToolbar = ReaderModeBarView(frame: CGRect.zero)
-        readerModeToolbar!.delegate = getApp().browserViewController
-        addSubview(readerModeToolbar!)
-        self.setNeedsLayout()
+        readerModeToolbar.isHidden = false
```
Comment (jhreis) [inline brave/src/frontend/BraveURLBarView.swift]: I know these are used outside of your changes, but I think we can just remove these now that they are single line implementations. Makes setting the hidden property much easier.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift (hunk 1 above)
Comment (iccub) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift]: Got it, just saw somewhere in code this if !BRAVE and parroted it

Ok, should I remove the whole method and its occurrences then?

Comment (jhreis) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+ReaderModeDelegate.swift]: Yeah, we used to be more careful about this, but not anymore. Again, no way for you to know.

Yes, if it is not needed, go ahead and destroy it 💣

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift [hunk 6]
```diff
@@ -554,7 +554,7 @@ class BrowserViewController: UIViewController {
 
         readerModeBar?.snp.remakeConstraints { make in
             make.top.equalTo(self.header.snp.bottom).constraint
-            make.height.equalTo(BraveUX.ReaderModeBarHeight)
+            make.height.equalTo(UIConstants.ToolbarHeight)
```
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift]: Curious about change away from `BraveUX`

File: brave/src/frontend/BraveUX.swift [hunk 7]
```diff
@@ -12,7 +12,7 @@ struct BraveUX {
 
     static var PanelShadowWidth = 11
 
-    static let ReaderModeBarHeight = 28
+    static let ReaderModeBarHeight = 24
```
Comment (jhreis) [inline brave/src/frontend/BraveUX.swift]: This is getting pretty small. Do you feel this just looks better?

Comment (iccub) [reply brave/src/frontend/BraveUX.swift]: I did't change it, number 24 was hardcoded in BraveURLBarView previously. This constant was used on Firefox reader bar, by mistake I guess

Comment (jamesmudgett) [reply brave/src/frontend/BraveUX.swift]: Would prefer to keep height larger size or go even taller than 28px.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift (hunk 6 above)
Comment (jamesmudgett) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift]: This doesn't seem intentional and may be reason for adjustment below. Height change below is likely not causing any changes in testing.

Should keep this on BraveUX, and tweak height as you did but see comments below.

Comment (iccub) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift]: Actually I could remove this code as it's no longer used, it's the old firefox reader bar. Brave reader bar is created in BraveURLBarView

File: brave/src/frontend/BraveUX.swift (hunk 7 above)
Comment (jhreis) [reply brave/src/frontend/BraveUX.swift]: @iccub Gotcha, yeah we were using our own value, 28 previously. Will let @jamesmudgett make the call on the final height (i.e. 28 or some other larger value)

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift [hunk 8]
```diff
@@ -552,12 +548,6 @@ class BrowserViewController: UIViewController {
             make.height.equalTo(BrowserViewControllerUX.ShowHeaderTapAreaHeight)
         }
 
-        readerModeBar?.snp.remakeConstraints { make in
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController.swift]: Removed it alltogether, Brave reader bar still works so this code is not needed, and just confuses ppl

File: brave/src/frontend/BraveUX.swift (hunk 7 above)
Comment (jamesmudgett) [reply brave/src/frontend/BraveUX.swift]: @iccub lets keep the height 28, we can make this design update at a later time.Than you, the rest looks great!

Comment (iccub) [reply brave/src/frontend/BraveUX.swift]: ok done, np

**Pull request merged**